### ISSUE TITLE
Struct arg "reinterpretation" leads to asserts in genPutArgStk on ARM32/64

### ISSUE THREAD

**Issue by mikedn** [dotnet/coreclr#23794]
```C#
using System;
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;

class Program
{
    [StructLayout(LayoutKind.Sequential)]
    struct S
    {
        public int i0;
        public int i1;
        public int i2;
        public int i3;

        public int i4;
        public int i5;
    }

    [StructLayout(LayoutKind.Sequential)]
    struct S16
    {
        public int i0;
        public int i1;
        public int i2;
        public int i3;
    }

    static void Main() => Test(default);

    [MethodImpl(MethodImplOptions.NoInlining)]
    static void Call(int r0, int r1, int r2, int r3, int r4, int r5, int r6, S16 s)
    {
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    static void Escape<T>(ref T t)
    {
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    static void Test(S ss)
    {
        var s = ss;
        Escape(ref s);
        Call(0, 1, 2, 3, 4, 5, 6, Unsafe.As<S, S16>(ref s));
    }
}
```
ARM32:
```
Assert failure(PID 14560 [0x000038e0], Thread: 1428 [0x0594]): Assertion failed '!"Bad value of 'gcType'"' in 'Program:Test(struct)' (IL size 34)

    File: d:\projects\coreclr\src\jit\compiler.cpp Line: 467
    Image: D:\Projects\coreclr\bin\Product\Windows_NT.x86.Checked\CoreRun.exe
```
ARM64:
```
Assert failure(PID 1748 [0x000006d4], Thread: 9912 [0x26b8]): Assertion failed 'structSize <= 2 * TARGET_POINTER_SIZE' in 'Program:Test(struct)' (IL size 34)

    File: d:\projects\coreclr\src\jit\codegenarmarch.cpp Line: 786
    Image: D:\Projects\coreclr\bin\Product\Windows_NT.x64.Checked\CoreRun.exe
```

**Comment by mikedn** [comment]
Both asserts are "noway" so probably they'll turn into `InvalidProgramException` at runtime. However, the ARM32 is caused by an out of bounds array read so the behavior may be random. If you're (un)lucky and the memory happens to contain a valid GC type then the result is likely bad codegen.

This all has to do with `genPutArgStk` attempting to retrieve information from the lclvars for `OBJ(LCL_VAR_ADDR)` trees. This can be safely done only when the OBJ and the lclvar have the same type.

**Comment by mikedn** [comment]
To add to the confusion, the actual copying code is trying to handle sizes that are not multiple of `TARGET_POINTER_SIZE` but that's not needed because morph handles such cases by copying the struct arg to a local (and then codegen uses the round up size of the local). Well, morph shouldn't really do that but then codegen's handling of struct sizes like 7 is broken/incomplete.

**Comment by mikedn** [comment]
@dotnet/jit-contrib Do you think this is something that should be fixed for 3.0? This kind of struct reinterpretation is a bit of an odd case in the .NET world and I wouldn't be surprised that there are other issues in this area of the JIT.

I'd be happy to work on a fix but I would sort of prefer to do that after #21705 is done. That encapsulates struct layout information such as size and GC layout in a class that makes it easier and safer to deal with. IMO the current state of affairs is what leads to bugs like this one.

**Comment by CarolEidt** [comment]
IMO this should be fixed for 3.0.

> I'd be happy to work on a fix but I would sort of prefer to do that after #21705 is done.

I don't think we can take #21705 for 3.0, but I'd love to see it done soon! 

> IMO the current state of affairs is what leads to bugs like this one.

Agree

**Comment by CarolEidt** [comment]
@dotnet/jit-contrib it would be great if someone could take this one; I want to focus on getting #16021 and #16022

**Comment by mikedn** [comment]
> IMO this should be fixed for 3.0.

OK then, I'll work on a minimal fix for this that doesn't require #21705. Thanks for feedback!
